scipy/sparse/csgraph/_flow.pyx

```diff
@@ -670,3 +683,943 @@ cdef ITYPE_t[:] _dinic(
+        return edge_flow[i]
+
+cdef edge_result _find_leaving_edge(
+        ITYPE_t[:] Wn,  # IN
+        ITYPE_t[:] We,  # IN
+        ITYPE_t[:] Wne_len,  # IN
+        ITYPE_t[:] edge_sources,  # IN
+        ITYPE_t[:] edge_targets,  # IN
+        ITYPE_t[:] edge_capacities,  # IN
+        ITYPE_t[:] edge_flow,  # IN
+        edge_result ret_values  # IN/OUT
+        ) nogil:
+    """
+    Returns the leaving edge in a cycle represented by Wn and We.
+    """
+    cdef:
+        ITYPE_t min_res_cap, res_cap
+        ITYPE_t i = -2, p, j = -2, s = -2, t, idx_e, idx_n
+    min_res_cap = INT_MAX
+    idx_n = Wne_len[0] - 1
+    idx_e = Wne_len[1] - 1
+    while idx_n >= 0 and idx_e >= 0:
+        i = We[idx_e]
+        p = Wn[idx_n]
+        res_cap = _residual_capacity(i, p,
+                                     edge_sources,
+                                     edge_capacities,
+                                     edge_flow)
+        if res_cap < min_res_cap:
+            min_res_cap = res_cap
+            j = i
+            s = p
+        idx_n -= 1
+        idx_e -= 1
+    if edge_sources[j] == s:
+        t = edge_targets[j]
+    else:
+        t = edge_sources[j]
+    ret_values.i = j
+    ret_values.p = s
+    ret_values.q = t
+    return ret_values
+
+cdef inline ITYPE_t _index(
+        ITYPE_t[:] W,  # IN
+        ITYPE_t W_len,  # IN
+        ITYPE_t i  # IN
+        ) nogil:
+    """
+    Returns the index of value in the given array.
+    """
+    cdef ITYPE_t idx
+    for idx in range(W_len):
+        if W[idx] == i:
+            return idx
+    return -1
+
+cdef inline void _augment_flow(
+        ITYPE_t[:] Wn,  # IN
+        ITYPE_t[:] We,  # IN
+        ITYPE_t[:] Wne_len,  # IN
+        ITYPE_t f,  # IN
+        ITYPE_t[:] edge_sources,  # IN
+        ITYPE_t[:] edge_flow  # IN/OUT
+        ) nogil:
+    """
+    Augment f units of flow along a cycle represented by Wn and We.
+    """
+    cdef ITYPE_t i, p, idx_e, idx_n
+    idx_e = 0
+    idx_n = 0
+    while (idx_n < Wne_len[0] and
+           idx_e < Wne_len[1]):
+        i = We[idx_n]
+        p = Wn[idx_e]
+        if edge_sources[i] == p:
+            edge_flow[i] += f
+        else:
+            edge_flow[i] -= f
+        idx_e += 1
+        idx_n += 1
+
+cdef void _remove_edge(
+        ITYPE_t s,  # IN/OUT
+        ITYPE_t t,  # IN/OUT
+        ITYPE_t[:] subtree_size,  # IN/OUT
+        ITYPE_t[:] prev_vertex_dft,  # IN/OUT
+        ITYPE_t[:] last_descendent_dft,  # IN/OUT
+        ITYPE_t[:] next_vertex_dft,  # IN/OUT
+        ITYPE_t[:] parent,  # IN/OUT
+        ITYPE_t[:] parent_edge,  # IN/OUT
+        ITYPE_t n_verts  # IN/OUT
+        ) nogil:
+    """
+    Remove an edge (s, t) where parent[t] == s from the spanning tree.
+    """
+    cdef:
+        ITYPE_t size_t, prev_t, last_t
+        ITYPE_t next_last_t
+    size_t = subtree_size[t]
+    prev_t = prev_vertex_dft[t]
+    last_t = last_descendent_dft[t]
+    next_last_t = next_vertex_dft[last_t]
```

Comment:
Initialization is generally better than definition and assignation.
Also [constant initialization](https://en.cppreference.com/w/cpp/language/constant_initialization) is a good practice.

Could this be the case here? Namely:
```suggestion
        const ITYPE_t size_t = subtree_size[t]
        const ITYPE_t prev_t = prev_vertex_dft[t]
        const ITYPE_t last_t = last_descendent_dft[t]
        const ITYPE_t next_last_t = next_vertex_dft[last_t]
```

If this is possible, `const` qualifying the argument of this function also might also make sense.